### PR TITLE
Add equiv scorer for matching descriptions

### PR DESCRIPTION
**Pull request by AdamDM**
- New scorer matches description to description
- Helps matching content with missing fields
- Potentially useful for sport where titles often mismatch

### CODE REVIEW

File: src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+
+public class DescriptionMatchingScorerTest {
+
+    private DescriptionMatchingScorer scorer;
+
+    public DescriptionMatchingScorerTest() {
+        this.scorer = DescriptionMatchingScorer.makeScorer();
```
Comment (jamiesprax) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: You should use the

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    private Set<String> descriptionToProcessedList(String description) {
+        return !Strings.isNullOrEmpty(description) ?
+               tokenize(description).stream()
+                .filter(s -> !commonWords.contains(s))
+                .collect(Collectors.toSet()) :
+                Sets.newHashSet();
+    }
+
+    private List<String> tokenize(String target) {
+        Pattern pattern = Pattern.compile("\\b([A-Z]\\w*)\\b");
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: Extract this string

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        //calibrate here
+        return new Double(candidateList.size())/new Double(subjectList.size()) > PROPORTION_CROSSOVER;
+    }
+
+    private Set<String> descriptionToProcessedList(String description) {
+        return !Strings.isNullOrEmpty(description) ?
+               tokenize(description).stream()
+                .filter(s -> !commonWords.contains(s))
+                .collect(Collectors.toSet()) :
+                Sets.newHashSet();
```
Comment (jamiesprax) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: This is a little confusing to read as the y part of a `x ? y : z` is what usually happens when x expression is true, but you've negated the expression so now they're swapped. Without noticing the `!` it looks like y happens when `description` is null or empty

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+public class DescriptionMatchingScorer implements EquivalenceScorer<Item> {
+
+    public static final String NAME = "Description Matching";
+    private static final Set<String> commonWords = ImmutableSet.of(
+            "the", "in", "a", "and", "&", "of", "to", "show"
+    );
+    //proportion crossover between key words in two descriptions threshold
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: This is not very easy to follow. Maybe 

```
// Proportion threshold of key words that need to match between the two descriptions
// to conclude there is a match

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    public static final double PROPORTION_CROSSOVER = 0.2;
+
+    private DescriptionMatchingScorer() {
+    }
+
+    public static DescriptionMatchingScorer makeScorer(){
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: :+1:

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        return descriptionMatch(subject, candidate) ? Score.ONE : Score.nullScore();
+    }
+
+    private boolean descriptionMatch(Item subject, Item candidate) {
+        Set<String> candidateList = descriptionToProcessedList(candidate.getDescription());
+        Set<String> subjectList = descriptionToProcessedList(subject.getDescription());
+
+        candidateList.retainAll(subjectList);
+
+        //calibrate here
```
Comment (Dalamar42) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: This is not very clear. Is this a TODO to calibrate in the future or does this code do the calibration. Also instead of calibration I think you mean normalise

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java (hunk 3 above)
Comment (Dalamar42) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: This is hard to follow.

- As @jamiesprax in if statements and especially in ternary operators positive conditions are generally preferred over negative conditions as they are easier to read and comprehend.

- With long ternary operators the code style is:
```
condition
? if branch
: else branch
```
however...

- This is definitely too much code to use a ternary operator. A standard if-else would be much more readable here.

```
if (condition) {
    return one thing;
} else {
    return another thing;
}
```

File: src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java (hunk 2 above)
Comment (Dalamar42) [inline src/main/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorer.java]: That ^ and give a helpful name of what it is this regex is supposed to match

File: src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java (hunk 1 above)
Comment (Dalamar42) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: Tests are not initialised in their constructor. There shouldn't be a constructor at all here.
Instead:

- For things that need to happen once only you use `@BeforeClass`. This is rare and usually only used to do expensive work like initialise a database
- For things that need to happen before every test like creating this scorer you use `@Before`

There is a code style on how these methods need to be named. Check the Junit documentation.

File: src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    private Score score(Item subject, Item candidate) {
+        DefaultDescription desc = new DefaultDescription();
+        ScoredCandidates<Item> scores = scorer.score(subject, ImmutableSet.of(candidate), desc);
+        return scores.candidates().get(candidate);
+    }
+
+    @Test
+    public void testDescriptionsMatchWhenTheyShould() {
```
Comment (Dalamar42) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: don't prefix tests with `test`. That's the old Junit style from the time of Junit3 that discovered tests by looking for methods that started with `test`. It's not needed any more

Comment (Dalamar42) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: It would be good to name this more descriptively. When should the descriptions match? `Should` is not obvious enough for someone unfamiliar with the code

File: src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        Item subject = new Item();
+        Item candidate = new Item();
+        subject.setDescription("Football related content and Manchester United talk");
+        candidate.setDescription("Manchester united play Football");
+        assertEquals(Score.ONE, score(candidate, subject));
+        assertEquals(Score.ONE, score(subject, candidate));
+    }
+
+    @Test
+    public void testDescriptionsMismatchWhenTheyShould() {
```
Comment (Dalamar42) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: ditto

File: src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        Item subject = new Item();
+        Item candidate = new Item();
+        subject.setDescription("Football related content and Manchester United talk");
+        candidate.setDescription("Something nothing Actually related");
+        assertEquals(Score.nullScore(), score(candidate, subject));
+        assertEquals(Score.nullScore(), score(subject, candidate));
+    }
+
+    @Test
+    public void testCapitalsAreNecessaryForMatching() {
```
Comment (Dalamar42) [inline src/test/java/org/atlasapi/equiv/scorers/DescriptionMatchingScorerTest.java]: This is a good name. You just need to remove the `test` bit
